pkg/mic/mic.go

```diff
@@ -874,31 +865,72 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
 // cleanUpAllAssignedIdentitiesOnNode deletes all assigned identities associated with a the node
 func (c *Client) cleanUpAllAssignedIdentitiesOnNode(node string, nodeTrackList trackUserAssignedMSIIds) {
 	glog.Infof("deleting all assigned identites for node %s", node)
-	for _, createID := range nodeTrackList.assignedIDsToCreate {
-		binding := createID.Spec.AzureBindingRef
+	for _, deleteID := range nodeTrackList.assignedIDsToDelete {
+		binding := deleteID.Spec.AzureBindingRef
 
-		err := c.removeAssignedIdentity(&createID)
+		err := c.removeAssignedIdentity(&deleteID)
 		if err != nil {
 			c.EventRecorder.Event(binding, corev1.EventTypeWarning, "binding remove error",
-				fmt.Sprintf("Removing assigned identity binding %s node %s for pod %s resulted in error %v", binding.Name, createID.Spec.NodeName, createID.Name, err.Error()))
+				fmt.Sprintf("Removing assigned identity binding %s node %s for pod %s resulted in error %v", binding.Name, deleteID.Spec.NodeName, deleteID.Name, err.Error()))
 			glog.Error(err)
 			continue
 		}
 		c.EventRecorder.Event(binding, corev1.EventTypeNormal, "binding removed",
-			fmt.Sprintf("Binding %s removed from node %s for pod %s", binding.Name, createID.Spec.NodeName, createID.Spec.Pod))
+			fmt.Sprintf("Binding %s removed from node %s for pod %s", binding.Name, deleteID.Spec.NodeName, deleteID.Spec.Pod))
 	}
+}
 
-	for _, deleteID := range nodeTrackList.assignedIDsToDelete {
-		binding := deleteID.Spec.AzureBindingRef
+func (c *Client) consolidateVMSSNodes(nodeMap map[string]trackUserAssignedMSIIds) {
+	vmssMap := make(map[string][]string)
 
-		err := c.removeAssignedIdentity(&deleteID)
+	for nodeName, nodeTrackList := range nodeMap {
+		node, err := c.NodeClient.Get(nodeName)
+		if err != nil && !strings.Contains(err.Error(), "not found") {
+			glog.Errorf("Unable to get node %s. Error %v", nodeName, err)
+			continue
+		}
+		if err != nil && strings.Contains(err.Error(), "not found") {
+			glog.Warningf("Unable to get node %s while updating user msis. Error %v", nodeName, err)
+
+			// node is no longer found in the cluster, all the assigned identities that were created in this sync loop
+			// and those that already exist for this node need to be deleted.
+			go c.cleanUpAllAssignedIdentitiesOnNode(nodeName, nodeTrackList)
```

Comment:
Is there a reason for us to perform this async ?

Reply:
Is there anything preventing another sync loop to come and perform the same operations again ?

Reply:
We don't want the clean up to block the node processing for rest of the nodes/identities. Good point about preventing sync loop doing again. We can create a WaitGroup at sync loop and pass that for consolidating (used if there is a clean up for node not existing) and for update node identities. That way we can block until all node processing is complete.